src/ModelContextProtocol/Protocol/Transport/StreamableHttpClientSessionTransport.cs

```diff
@@ -0,0 +1,267 @@
+        SetConnected(true);
+    }
+
+    /// <inheritdoc/>
+    public override async Task SendMessageAsync(
+        JsonRpcMessage message,
+        CancellationToken cancellationToken = default)
+    {
+        using var sendCts = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken, _connectionCts.Token);
+        cancellationToken = sendCts.Token;
+
+        using var content = new StringContent(
+            JsonSerializer.Serialize(message, McpJsonUtilities.JsonContext.Default.JsonRpcMessage),
+            Encoding.UTF8,
+            "application/json"
+        );
+
+        using var httpRequestMessage = new HttpRequestMessage(HttpMethod.Post, _options.Endpoint)
+        {
+            Content = content,
+            Headers =
+            {
+                Accept = { ApplicationJsonMediaType, TextEventStreamMediaType },
+            },
+        };
+
+        CopyAdditionalHeaders(httpRequestMessage.Headers, _options.AdditionalHeaders, _mcpSessionId);
+        var response = await _httpClient.SendAsync(httpRequestMessage, HttpCompletionOption.ResponseHeadersRead, cancellationToken).ConfigureAwait(false);
+
+        response.EnsureSuccessStatusCode();
+
+        var rpcRequest = message as JsonRpcRequest;
+        JsonRpcMessage? rpcResponseCandidate = null;
+
+        if (response.Content.Headers.ContentType?.MediaType == "application/json")
+        {
+            var responseContent = await response.Content.ReadAsStringAsync(cancellationToken).ConfigureAwait(false);
+            if (responseContent.StartsWith("["))
+            {
+                rpcResponseCandidate = await ProcessJsonRpcBatch(responseContent, rpcRequest, cancellationToken).ConfigureAwait(false);
+            }
+            else
+            {
+                rpcResponseCandidate = await ProcessMessageAsync(responseContent, cancellationToken).ConfigureAwait(false);
+            }
+        }
+        else if (response.Content.Headers.ContentType?.MediaType == "text/event-stream")
+        {
+            var responseBodyStream = await response.Content.ReadAsStreamAsync(cancellationToken);
+            rpcResponseCandidate = await ProcessSseResponseAsync(responseBodyStream, rpcRequest, cancellationToken).ConfigureAwait(false);
+        }
+
+        if (rpcRequest is null)
+        {
+            return;
+        }
+
+        if (rpcResponseCandidate is not JsonRpcMessageWithId messageWithId || messageWithId.Id != rpcRequest.Id)
+        {
+            throw new McpException($"Streamable HTTP POST response completed without a reply to request with ID: {rpcRequest.Id}");
+        }
+
+        if (rpcRequest.Method == RequestMethods.Initialize && rpcResponseCandidate is JsonRpcResponse)
+        {
+            // We've successfully initialized! Copy session-id and start GET request.
+            _mcpSessionId = response.Headers.GetValues("mcp-session-id").Single();
```

Comment:
Is this guaranteed to exist? Wondering if we need a better exception than Single will produce if it doesn't.